README.md
# Taxonomies Sortable

Taxonomies Sortable is WordPress plugin that allow to sort non-hierarchical taxonomy terms associated to objects directly by drag&drop them on the tags metaboxes administration interface.

# Installation

This section describes how to install the plugin and get it working.

1. Upload the plugin files to the `/wp-content/plugins/taxonomies-sortable` directory, or install the plugin through the WordPress plugins screen directly.
1. Activate the plugin through the _Plugins_ screen in WordPress.

# Usage

Once the plugin is installed you can make a taxonomy sortable in the following ways:

* Programmatically, by adding `'sortable' => true` argument in `register_taxonomy()`.
* Programmatically, by using `taxonomies_sortable` filter below.
* Using the _Settings->Taxonomies Sortable_ administration screen by selecting non-hierarchical taxonomies you want to make sortable.

Note that making a taxonomy sortable, implicitly means also setting `sort => true` for that taxonomy.

# Hooks

## `taxonomies_sortable`

Filter allowing to programmatically set sortable taxonomies.

```php
apply_filters( 'taxonomies_sortable', array $taxonomies )
```

## `taxonomies_sortable_admin_settings`

Filter allowing to display or not the plugin settings page in the administration.

```php
apply_filters( 'taxonomies_sortable_admin_settings', boolean $display )
```

# Frequently Asked Questions

## Does it work with Gutenberg?

Currently no, even if the Gutenberg support is under development.

# Screenshots 

### Admin tag metabox ###

The taxonomy terms sorting drag&drop in action.

![Admin metabox](https://raw.githubusercontent.com/enrico-sorcinelli/taxonomies-sortable/master/assets-wp/screenshot-1.png)

### Plugin settings ###

The plugin settings page.

![Plugin settings](https://raw.githubusercontent.com/enrico-sorcinelli/taxonomies-sortable/master/assets-wp/screenshot-2.png)

# License: GPLv2 #

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.1] - 2018-06-22
### Fixed
 - Uninstall plugin hook.
 - Fixed static method declaration.
 - `TAXONOMIES_SORTABLE_PLUGIN_AUTOENABLE` constant typo and added documentation.

### Changed
 - Doc improvements.

## [1.0.0] - 2018-06-20
### Changed
 - Version number for stable release.

### Fixed
 - Force `wp_get_object_terms()` loop for taxonomies with differents `$args`. 
 - PHPCS fixes. 

## [0.0.1] - 2017-08-14
### Added
 - Initial release.


assets/js/admin.js
/**
 * @file This file contains Taxonomies Sortable JavaScript
 * @author Enrico Sorcinelli
 * @version 1.0.1
 * @title Taxonomies Sortable
 */

// Make all in a closure.
;( function ( $ ) {

	$( document ).ready( function() {

		// Post screen.
		var tagDelimiter = ( window.tagsSuggestL10n && window.tagsSuggestL10n.tagDelimiter ) || ',';

		// Add sortable UI.
		if ( 'object' === typeof( taxonomies_sortable_plugin_i18n.taxonomies ) ) {
			$( taxonomies_sortable_plugin_i18n.taxonomies.join( ',' ) ).each( function() {
				var $wrapper = $( this ).find( '.tagchecklist' ),
					$the_tags = $( this ).find( '.the-tags' );
				$wrapper.sortable( {
					items: '>li',
					stop: function ( event, ui ) {
						var new_tags = [];
						$( 'li', $wrapper ).each( function( i ) {
							var $button = $( 'button.ntdelbutton', this );
							$button.attr( 'id', $button.attr( 'id' ).replace( /\d+$/, i ) );
							new_tags.push( $( $( this ).contents()[2] ).text() );
						} );
						$the_tags.val( array_unique_noempty( new_tags ).join( tagDelimiter ) );
					}
				} );
			} );
		}

		// Tab managements.
		$( document ).ready( function() {
			$( '.nav-tab-wrapper.taxonomies-sortable a' ).on( 'click', function () {
				$( '.nav-tab.taxonomies-sortable' ).removeClass( 'nav-tab-active' );
				$( this ).addClass( 'nav-tab-active' );
				$( 'section.taxonomies-sortable' ).hide();
				$( 'section.taxonomies-sortable' ).eq( $( this ).index() ).show();
				return false;
			});
		})

	})

} )( jQuery );


taxonomies-sortable.php
<?php
/**
 * Plugin Taxonomies Sortable
 *
 * @package taxonomies-sortable
 */

/**
 * Plugin Name: Taxonomies Sortable
 * Plugin URI:  https://github.com/enrico-sorcinelli/taxonomies-sortable
 * Description: A WordPress Taxonomies Sortable
 * Author:      Enrico Sorcinelli
 * Author URI:  https://github.com/enrico-sorcinelli/taxonomies-sortable/graphs/contributors
 * Text Domain: taxonomies-sortable
 * Domain Path: /languages/
 * Version:     1.0.1
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

if ( ! class_exists( 'Taxonomies_Sortable' ) ) {

	// Plugins constants.
	define( 'TAXONOMIES_SORTABLE_PLUGIN_VERSION', '1.0.1' );
	define( 'TAXONOMIES_SORTABLE_PLUGIN_BASEDIR', dirname( __FILE__ ) );
	define( 'TAXONOMIES_SORTABLE_PLUGIN_BASEURL', plugin_dir_url( __FILE__ ) );

	// Enable debug prints on error_log (only when WP_DEBUG is true).
	if ( ! defined( 'TAXONOMIES_SORTABLE_PLUGIN_DEBUG' ) ) {
		define( 'TAXONOMIES_SORTABLE_PLUGIN_DEBUG', false );
	}

	require_once TAXONOMIES_SORTABLE_PLUGIN_BASEDIR . '/php/class-taxonomies-sortable.php';

	/**
	 * Init the plugin.
	 *
	 * Define TAXONOMIES_SORTABLE_PLUGIN_AUTOENABLE to `false` in your
	 * _wp-config.php_ to disable.
	 *
	 * @return void
	 */
	function taxonomies_sortable_plugin_init() {

		if ( defined( 'TAXONOMIES_SORTABLE_PLUGIN_AUTOENABLE' ) && TAXONOMIES_SORTABLE_PLUGIN_AUTOENABLE === false ) {
			return;
		}

		// Instantiate plugin class and add it to the set of globals.
		$GLOBALS['taxonomies_sortable_plugin'] = Taxonomies_Sortable::get_instance( array( 'debug' => TAXONOMIES_SORTABLE_PLUGIN_DEBUG && WP_DEBUG ) );
	}

	// Activate the plugin once all plugin have been loaded.
	add_action( 'plugins_loaded', 'taxonomies_sortable_plugin_init' );

	// Uninstall hooks.
	register_uninstall_hook( __FILE__, array( 'Taxonomies_Sortable', 'pluginUninstall' ) );
}


php/adminpages/credits.php
<?php
/**
 * Credit template part.
 *
 * @package taxonomies-sortable
 */

?>
			<div id="postbox-container-1" class="postbox-container">
				<div class="postbox" id="taxonomies-sortable-credits">
					<h3 class="hndle"><?php /* translators: */ echo esc_html( sprintf( __( 'Taxonomies Sortable %s', 'taxonomies-sortable' ), TAXONOMIES_SORTABLE_PLUGIN_VERSION ) ); ?></h3>
					<div class="inside">
						<h4><?php esc_html_e( 'Changelog', 'taxonomies-sortable' ); ?></h4>
						<p>
							<?php esc_html_e( 'What\'s new in', 'taxonomies-sortable' ); ?>
							<a href="https://github.com/enrico-sorcinelli/taxonomies-sortable/releases"><?php /* translators: */ echo esc_html( sprintf( __( 'version %s', 'taxonomies-sortable' ), TAXONOMIES_SORTABLE_PLUGIN_VERSION ) ); ?></a>.
						</p>
						<h4><?php esc_html_e( 'Support', 'taxonomies-sortable' ); ?></h4>
						<p>
							<span class="dashicons dashicons-email-alt"></span>
							<a href="https://github.com/enrico-sorcinelli/taxonomies-sortable/issues"><?php esc_html_e( 'Support/issues', 'taxonomies-sortable' ); ?></a>.
						</p>
						<div class="author">
							<i><span><?php esc_html_e( 'Taxonomies Sortable', 'taxonomies-sortable' ); ?> <?php esc_html_e( 'by', 'taxonomies-sortable' ); ?> Enrico Sorcinelli</span><br>
							&copy; <?php echo esc_html( date( __( 'Y', 'taxonomies-sortable' ) ) ); ?></i>
						</div>
					</div>
				</div>
			</div>


php/class-plugin-utils.php
<?php
/**
 * Plugin utilities factory class.
 *
 * @package taxonomies-sortable
 * @author Enrico Sorcinelli
 */

/**
 * Class Plugin_Utils
 */
class Plugin_Utils {

	/**
	 * Check for AJAX request.
	 *
	 * @return boolean
	 */
	public static function isAjaxRequest() {

		if (
			( ! empty( $_SERVER['HTTP_X_REQUESTED_WITH'] ) && strtolower( $_SERVER['HTTP_X_REQUESTED_WITH'] ) === 'xmlhttprequest' )
			|| ( defined( 'DOING_AJAX' ) && DOING_AJAX )
		) {
				return true;
		}

		return false;
	}

	/**
	 * Call a template. Return also value returned from `include`.
	 *
	 * @param string|array $template Path of file to include (include_path),
	 *                               function or class method.
	 * @param array|null   $params   Associative array that you could use
	 *                               inside template. Use `$params{key}` in the
	 *                               template to refer to the specific `key`.
	 *
	 * @return mixed
	 */
	public static function includeTemplate( $template, $params = array() ) {

		// WP globals.
		global $wp_query, $wp_style, $wp_registered_sidebars, $sidebars_widgets, $wp_registered_widgets, $posts, $post, $wp_did_header, $wp_rewrite, $wpdb, $wp_version, $wp, $id, $comment, $user_ID;

		// Check for a callable template (function/class method).
		if ( is_callable( $template ) ) {
			return call_user_func( $template, $params );
		}

		// Consider template a file.
		return include( $template );
	}

	/**
	 * Call a template by buffering output (like `sprintf`).
	 *
	 * @param string $template Path of file to include (include_path).
	 * @param array  $params   Hash array that you could use inside template.
	 *                         Use `$params{key}` in access to refer to the
	 *                         specific `key`.
	 *
	 * @return mixed Return all standard output generated by template or the
	 *               value returned by template if it doesn't return a scalar
	 *               (i.e an object).
	 */
	public static function sincludeTemplate( $template, $params = array() ) {

		// WP globals.
		global $wp_query, $wp_style, $wp_registered_sidebars, $sidebars_widgets, $wp_registered_widgets, $posts, $post, $wp_did_header, $wp_rewrite, $wpdb, $wp_version, $wp, $id, $comment, $user_ID;

		ob_start();

		// Check for a callable template (function/class method).
		if ( is_callable( $template ) ) {
			$ret = call_user_func( $template, $params );
		}
		else {
			$ret = include( $template );
		}

		$stdout = ob_get_clean();

		// Include fails.
		if ( false === $ret ) {
			return $ret;
		}

		// Checks return type.
		return ( is_scalar( $ret ) || null === $ret ) ? $stdout : $ret;
	}

	/**
	 * Returns true if `$regex` is a valid regular expression pattern.
	 *
	 * @param mixed $regex Regex to be tested.
	 *
	 * @return boolean
	 */
	public static function isRegex( $regex = null ) {

		// Pattern is broken.
		if ( @preg_match( $regex, null ) === false ) {
			return false;
		}

		// Pattern is valid.
		return true;
	}

}


php/adminpages/general-settings.php
<?php
/**
 * Plugin settings option page template.
 *
 * @package taxonomies-sortable
 */

// Get the registered taxonomies.
$registered_taxonomies = get_taxonomies( '', 'objects' );
?>
<div class="wrap">
<h1><?php esc_html_e( 'Taxonomies Sortable settings', 'taxonomies-sortable' ); ?></h1>

	<div id="poststuff">
		<div id="post-body" class="metabox-holder columns-2">

			<div id="postbox-container-2" class="postbox-container">
<form method="post" action="options.php">

<?php settings_fields( $params['prefix'] . 'general-settings' ); ?>
<?php do_settings_sections( $params['prefix'] . 'general-settings' ); ?>

	<h2 class="nav-tab-wrapper wp-clearfix taxonomies-sortable">
		<a class="nav-tab taxonomies-sortable nav-tab-active"><?php esc_html_e( 'Taxonomies', 'taxonomies-sortable' ); ?></a>
		<a class="nav-tab taxonomies-sortable"><?php esc_html_e( 'Options', 'taxonomies-sortable' ); ?></a>
	</h2>

<section class="taxonomies-sortable">
	<table class="form-table">
		<tr>
			<th scope="row"><?php esc_html_e( 'Taxonomies sortable', 'taxonomies-sortable' ); ?></th>
			<td>
				<fieldset id="sortable-taxonomies-fieldset">
<?php
if ( ! empty( $registered_taxonomies ) ) {
	foreach ( (array) $registered_taxonomies as $tax ) {

		// Skip hierarchical and builtin (except of post_tag).
		if ( $tax->hierarchical || ( $tax->_builtin && 'post_tag' !== $tax->name ) ) {
			continue;
		}
?>
					<label for="<?php echo esc_attr( $params['prefix'] . $tax->name ); ?>">
						<input class="taxonomies-sortable" id="<?php echo esc_attr( $params['prefix'] . $tax->name ); ?>" type="checkbox" name="<?php echo esc_attr( $params['prefix'] ); ?>general_settings[taxonomies][]" value="<?php echo esc_attr( $tax->name ); ?>" <?php checked( 1, ! empty( $params['settings']['taxonomies'] ) && in_array( $tax->name, $params['settings']['taxonomies'], true ), true ); ?> />
						<?php echo esc_html( $tax->labels->name ); ?>
					</label>
				<br>
<?php
	}
}
?>
				</fieldset>
				<p class="description"><?php esc_html_e( 'Select non-hierarchical taxonomies to make sortable.', 'taxonomies-sortable' ); ?></p>
			</td>
		<tr>
	</table>
</section>

<section class="taxonomies-sortable">
	<table class="form-table">
		<tr>
			<th scope="row"><?php esc_html_e( 'Plugin settings', 'taxonomies-sortable' ); ?></th>
			<td>
				<p>
					<input name="<?php echo esc_attr( $params['prefix'] . 'general_settings[remove_plugin_settings]' ); ?>" type="checkbox" id="<?php echo esc_attr( $params['prefix'] . 'general_settings_remove_plugin_settings' ); ?>" value="1" <?php checked( 1, empty( $params['settings']['remove_plugin_settings'] ) ? 0 : 1, true ); ?>>
					<label for="<?php echo esc_attr( $params['prefix'] . 'general_settings_remove_plugin_settings' ); ?>"><?php esc_html_e( 'Completely remove options on plugin removal.', 'taxonomies-sortable' ); ?></label>
				</p>
			</td>
		<tr>
	</table>
</section>

<?php submit_button(); ?>
</form>
			</div>
<?php \Plugin_Utils::includeTemplate( TAXONOMIES_SORTABLE_PLUGIN_BASEDIR . '/php/adminpages/credits.php', array( 'prefix' => $params['prefix'] ) ); ?>
		</div><!-- /#post-body -->
	</div><!-- /#poststuff -->
</div><!--/.wrap-->


php/class-taxonomies-sortable-admin.php
<?php
/**
 * Plugin administration class.
 *
 * @package taxonomies-sortable
 * @author Enrico Sorcinelli
 */

namespace Taxonomies_Sortable;

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}
/**
 * Admin interface class.
 *
 * @package taxonomies-sortable
 * @author Enrico Sorcinelli
 */
class Admin {

	/**
	 * Admin pages name.
	 *
	 * @var array
	 */
	private $admin_pages = array();

	/**
	 * Prefix.
	 *
	 * @var string
	 */
	private $prefix;

	/**
	 * Plugin options.
	 *
	 * @var array
	 */
	private $plugin_options;

	/**
	 * Construct the plugin.
	 *
	 * @param array $args {
	 *     Array of arguments for constructor.
	 *
	 *     @type string  $prefix
	 *     @type boolean $debug  Default `false`.
	 * }
	 *
	 * @return object
	 */
	public function __construct( $args = array() ) {

		// Set object property.
		$this->debug = isset( $args['debug'] ) ? $args['debug'] : false;
		foreach ( array( 'prefix', 'plugin_options' ) as $property ) {
			$this->$property = $args[ $property ];
		}

		// This plugin only runs in admin, but we need it initialized on init.
		add_action( 'init', array( $this, 'init' ) );
	}

	/**
	 * Initialize the plugin: setup menu, settings, add filters, actions,
	 * scripts, styles and so on.
	 *
	 * @return void
	 */
	public function init() {

		if ( ! is_admin() ) {
			return;
		}

		// Menu settings.
		add_action( 'admin_menu', array( $this, 'setupMenu' ), '10' );

		// Forms settings.
		add_action( 'admin_init', array( $this, 'setupSettings' ), '10' );
	}

	/**
	 * Setup admin menu.
	 *
	 * @return void
	 */
	public function setupMenu() {

		$admin_menu_capability = is_super_admin() ? 'manage_options' : 'taxonomies_sortable_plugin_manage_options';

		/**
		 * Filter to allow display plugin settings page.
		 *
		 * @param boolean $display_setting_page Default `true`.
		 *
		 * @return boolean
		 */
		if ( apply_filters( 'taxonomies_sortable_admin_settings', true ) ) {
			$this->admin_pages['general_settings'] = add_options_page( __( 'Taxonomies Sortable settings', 'taxonomies-sortable' ), __( 'Taxonomies Sortable', 'taxonomies-sortable' ), 'manage_options', $this->prefix . 'general_settings', array( $this, 'pageGeneralSettings' ) );
		}
	}

	/**
	 * Setup admin (register settings, enqueue ie JavaScript, CSS, add filters &
	 * actions, ...).
	 *
	 * @return void
	 */
	public function setupSettings() {

		global $wp_version;

		// Enqueue JS/CSS only for non AJAX requests.
		if ( ! \Plugin_Utils::isAjaxRequest() ) {

			// Screens where to enqueue assets.
			$admin_pages = array_merge( array( 'post.php', 'post-new.php' ), array_values( $this->admin_pages ) );

			// Add CSS to post pages.
			foreach ( $admin_pages as $page ) {
				add_action( 'admin_print_styles-' . $page, array( $this, 'loadCSS' ), 10, 0 );
				add_action( 'admin_print_scripts-' . $page, array( $this, 'loadJavaScript' ), 10, 0 );
			}
		}

		// Register settings.
		register_setting( $this->prefix . 'general-settings', $this->prefix . 'general_settings', array( $this, 'checkGeneralSettings' ) );
	}

	/**
	 * Check general settings.
	 *
	 * @param mixed $settings Settings values.
	 *
	 * @return array
	 */
	public function checkGeneralSettings( $settings ) {

		return $settings;
	}

	/**
	 * Load CSS files.
	 *
	 * @return void
	 */
	public function loadCSS() {

		global $post_type;

		wp_enqueue_style(
			$this->prefix . 'css',
			TAXONOMIES_SORTABLE_PLUGIN_BASEURL . 'assets/css/admin.css',
			array()
		);
	}

	/**
	 * Load JavaScript files.
	 *
	 * @return void
	 */
	public function loadJavaScript() {

		global $post_type;

		wp_enqueue_script(
			$this->prefix . 'js',
			TAXONOMIES_SORTABLE_PLUGIN_BASEURL . 'assets/js/admin.js',
			array(),
			TAXONOMIES_SORTABLE_PLUGIN_VERSION,
			false
		);

		// Get sortable taxonomies for current post type.
		$tax_objs = get_object_taxonomies( $post_type, 'objects' );

		$sortable_taxonomies = array();
		foreach ( $tax_objs as $taxonomy ) {
			if ( ! $taxonomy->hierarchical && isset( $taxonomy->sortable ) && $taxonomy->sortable ) {
				$sortable_taxonomies[] = '#' . $taxonomy->name . '.tagsdiv';
			}
		}

		// Localization.
		wp_localize_script( $this->prefix . 'js', $this->prefix . 'i18n', array(
			'_plugin_url' => TAXONOMIES_SORTABLE_PLUGIN_BASEURL,
			'msgs'        => array(),
			'taxonomies'  => $sortable_taxonomies,
		) );
	}

	/**
	 * General settings admin page callback.
	 *
	 * @return void
	 */
	public function pageGeneralSettings() {

		\Plugin_Utils::includeTemplate( TAXONOMIES_SORTABLE_PLUGIN_BASEDIR . '/php/adminpages/general-settings.php', array(
			'prefix'   => $this->prefix,
			'settings' => $this->plugin_options,
		) );
	}
}


php/class-taxonomies-sortable.php
<?php
/**
 * Plugin base class.
 *
 * @package taxonomies-sortable
 * @author Enrico Sorcinelli
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

/**
 * Class Taxonomies_Sortable.
 */
class Taxonomies_Sortable {

	/**
	 * Plugin constrctor arguments.
	 *
	 * @var array
	 */
	private $settings = array();

	/**
	 * Plugin options.
	 *
	 * @var array
	 */
	private $plugin_options;

	/**
	 * Prefix used for options and postmeta fields, DOM IDs and DB tables.
	 *
	 * @var string
	 */
	private static $prefix = 'taxonomies_sortable_plugin_';

	/**
	 * Singleton instance.
	 *
	 * @var object Instance of this class.
	 */
	public static $instance;

	/**
	 * Plugin class constructor.
	 *
	 * @param array $args {
	 *     Array of arguments for constructor.
	 *
	 *     @type boolean $debug Default `false`.
	 * }
	 *
	 *  @return object
	 */
	public function __construct( $args = array() ) {

		$this->settings = wp_parse_args( $args, array(
			'debug' => false,
		), $args );

		$this->plugin_options = $this->getPluginOptions();

		// Load plugin text domain.
		load_plugin_textdomain( 'taxonomies-sortable', false, dirname( plugin_basename( __FILE__ ) ) . '/../languages/' );

		// Check and load needed components.
		$this->requireComponents();

		// Create plugin admin object.
		if ( is_admin() ) {
			$this->_admin = new \Taxonomies_Sortable\Admin( array(
				'prefix'         => self::$prefix,
				'plugin_options' => $this->plugin_options,
				'debug'          => $this->settings['debug'],
			));
		}

		// Add sort filter for sortable taxonomies.
		add_filter( 'wp_get_object_terms_args', array( $this, 'wpGgetOobjectTermsArgsFilter' ), 10, 3 );

		// Make taxonomies sortable.
		add_filter( 'register_taxonomy_args', array( $this, 'makeTaxonomiesSortable' ), 10, 2 );
	}

	/**
	 * Get the singleton instance of this class.
	 *
	 * @param array $args Constructor arguments.
	 *
	 * @return object
	 */
	public static function get_instance( $args = array() ) {
		if ( ! ( self::$instance instanceof self ) ) {
			self::$instance = new self( $args );
		}
		return self::$instance;
	}

	/**
	 * This function will include core files before the theme's functions.php
	 * file has been excecuted.
	 *
	 * @return void
	 */
	public function requireComponents() {

		// Plugin classes.
		if ( ! class_exists( 'Plugin_Utils' ) ) {
			require_once TAXONOMIES_SORTABLE_PLUGIN_BASEDIR . '/php/class-plugin-utils.php';
		}
		if ( is_admin() ) {
			require_once TAXONOMIES_SORTABLE_PLUGIN_BASEDIR . '/php/class-taxonomies-sortable-admin.php';
		}
	}

	/**
	 * Filter adding sortable arguments when registering taxonomy.
	 *
	 * @param array  $args     Taxonomy registration arguments.
	 * @param string $taxonomy Taxonomy name.
	 *
	 * @return array
	 */
	public function makeTaxonomiesSortable( $args, $taxonomy ) {

		if ( in_array( $taxonomy, $this->plugin_options['taxonomies'], true ) ) {
			$args['sortable'] = true;
		}
		if ( isset( $args['sortable'] ) && true === $args['sortable'] ) {
			$args['sort'] = true;
		}
		else {
			$args['sortable'] = false;
		}

		// In order to force wp_get_object_terms() loop in wp-includes/taxonomy.php:1975.
		$args['args']['_tax'] = $taxonomy;

		return $args;
	}

	/**
	 * Filter applied to set term meta order (see https://core.trac.wordpress.org/ticket/35925).
	 *
	 * @param array   $args       Query arguments.
	 * @param integer $object_ids Object ID.
	 * @param array   $taxonomies Taxonomies.
	 *
	 * @return array
	 */
	public function wpGgetOobjectTermsArgsFilter( $args, $object_ids, $taxonomies ) {

		$t = get_taxonomy( $taxonomies[0] );
		if ( isset( $t->sortable ) && true === $t->sortable ) {
			$args['orderby'] = 'term_order';
		}
		else {
			unset( $args['orderby'] );
		}

		return $args;
	}

	/**
	 * Plugin uninstall hook.
	 * Delete wp_options entries only when plugin deactivated and deleted.
	 *
	 * @return void
	 */
	public static function pluginUninstall() {
		$options = get_option( self::$prefix . 'general_settings', true );
		if ( isset( $options['delete'] ) && $options['delete'] ) {
			delete_option( self::$prefix . 'general_settings' );
		}
	}

	/**
	 * Get plugin options settings.
	 *
	 * @return array
	 */
	private function getPluginOptions() {
		$settings = wp_parse_args(
			get_option( self::$prefix . 'general_settings', array() ),
			array(
				'taxonomies'             => array(),
				'remove_plugin_settings' => false,
			)
		);

		/**
		 * Filter taxonomies to make sortable.
		 *
		 * @param array $taxonomies Taxonomies sortable.
		 *
		 * @return array
		 */
		$settings['taxonomies'] = apply_filters( 'taxonomies_sortable', $settings['taxonomies'] );

		return $settings;
	}
}
